File: stepik-union/src/main/java/org/stepik/plugin/projectWizard/pycharm/PyCharmWizardStep.java
package org.stepik.plugin.projectWizard.pycharm;

import com.intellij.facet.ui.ValidationResult;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.project.Project;
import org.jetbrains.annotations.NotNull;
import org.stepik.api.objects.StudyObject;
import org.stepik.core.SupportedLanguages;
import org.stepik.core.stepik.StepikAuthManager;
import org.stepik.core.stepik.StepikAuthManagerListener;
import org.stepik.core.stepik.StepikAuthState;
import org.stepik.plugin.projectWizard.ui.ProjectSettingListener;
import org.stepik.plugin.projectWizard.ui.ProjectSettingsPanel;

import javax.swing.*;

import static org.stepik.core.stepik.StepikAuthManager.authentication;
import static org.stepik.core.stepik.StepikAuthState.AUTH;
import static org.stepik.core.stepik.StepikAuthState.NOT_AUTH;

class PyCharmWizardStep implements ProjectSettingListener, StepikAuthManagerListener {
    private final ValidationResult invalidCourse = new ValidationResult("Please, select a course");
    private final ValidationResult needLogin = new ValidationResult("Please, you must login");
    private final StepikPyProjectGenerator generator;
    private final ProjectSettingsPanel panel;
    private StepikAuthState authenticated = NOT_AUTH;

    PyCharmWizardStep(@NotNull StepikPyProjectGenerator generator, @NotNull Project project) {
        this.generator = generator;
        panel = new ProjectSettingsPanel(false);
        panel.addListener(this);
        authenticated = authentication(false);
        StepikAuthManager.addListener(this);
    }

    @NotNull
    StudyObject getSelectedStudyObject() {
        return panel.getSelectedStudyObject();
    }

    @NotNull
    JPanel getComponent() {
        return panel.getComponent();
    }

    @NotNull
    ValidationResult check() {
        StudyObject selectedStudyObject = panel.getSelectedStudyObject();

        if (selectedStudyObject.getId() == 0) {
            return invalidCourse;
        }

        if (authenticated != AUTH) {
            return needLogin;
        }
        return ValidationResult.OK;
    }

    @Override
    public void changed() {
        generator.fireStateChanged();
    }

    void updateStep() {
        panel.setLanguage(SupportedLanguages.PYTHON3);
        panel.updateStep();
    }

    void dispose() {
        panel.dispose();
    }

    @Override
    public void stateChanged(@NotNull StepikAuthState oldState, @NotNull StepikAuthState newState) {
        authenticated = newState;
        ApplicationManager.getApplication().invokeLater(generator::fireStateChanged);
    }
}

File: stepik-union/src/main/kotlin/org/stepik/plugin/projectWizard/pycharm/PyCharmWizardStep.kt
package org.stepik.plugin.projectWizard.pycharm

import com.intellij.facet.ui.ValidationResult
import com.intellij.openapi.application.ApplicationManager
import org.stepik.core.SupportedLanguages
import org.stepik.core.auth.StepikAuthManager
import org.stepik.core.auth.StepikAuthManager.authentication
import org.stepik.core.auth.StepikAuthManagerListener
import org.stepik.core.auth.StepikAuthState
import org.stepik.core.auth.StepikAuthState.AUTH
import org.stepik.plugin.projectWizard.ui.ProjectSettingListener
import org.stepik.plugin.projectWizard.ui.ProjectSettingsPanel

internal class PyCharmWizardStep(private val generator: StepikPyProjectGenerator) :
        ProjectSettingListener, StepikAuthManagerListener {
    private val invalidCourse = ValidationResult("Please, select a course")
    private val needLogin = ValidationResult("Please, you must login")
    private val panel: ProjectSettingsPanel = ProjectSettingsPanel(false)
    private var authenticated = authentication(false)

    val selectedStudyObject = panel.selectedStudyObject

    val component = panel.getComponent()

    init {
        panel.addListener(this)
        StepikAuthManager.addListener(this)
    }

    fun check(): ValidationResult {
        return when {
            panel.selectedStudyObject.id == 0L -> invalidCourse
            authenticated != AUTH -> needLogin
            else -> ValidationResult.OK
        }
    }

    override fun changed() {
        generator.fireStateChanged()
    }

    fun updateStep() {
        panel.language = SupportedLanguages.PYTHON3
        panel.updateStep()
    }

    fun dispose() {
        panel.dispose()
    }

    override fun stateChanged(oldState: StepikAuthState, newState: StepikAuthState) {
        authenticated = newState
        ApplicationManager.getApplication().invokeLater { generator.fireStateChanged() }
    }
}
